unixfs.ts

```diff
@@ -0,0 +1,484 @@
+ * replication. Such files in UnixFS are represented via `AdvancedFileLayout`.
+ * And files that fit into a single block are represented via `SimpleFileLayout`.
+ * 
+ * Please note: In some configurations files that fit a single block are not
+ * even encoded as UnixFS but rather as a `RawNode` blocks. However this type
+ * describes UinxFS File representation and not IPFS file representation which
+ * is why `RawNode` variant is not part of it.
+ */
+export type FileLayout =
+  | SimpleFileLayout
+  | AdvancedFileLayout
+
+
+/**
+ * Logacal representation of a directory that fits single block.
+ */
+export interface FlatDirectoryLayout extends PBNode {
+  Data: ByteView<{
+    Type: DataType.Directory
+    /**
+     * Directories MUST have `filesize` set to `0`.
+     */
+    filesize: 0
+    /**
+     * If omitted to be interpreted as default `0755`. It is RECOMMENDED
+     * to omit if mode matches default.
+     */
+    mode?: Mode
+    mtime?: UnixTime
+  }>
+  /**
+   * Links are directory entries.
+   */
+  Links: NamedLink<FileLayout|DirectoryLayout>[]
+}
+
+
+/**
+ * Logical representation of directory encoded in multiple blocks (usually when
+ * it contains large number of entries). Such directories are represented via
+ * Hash Array Map Tries (HAMT).
+ * 
+ * @see https://en.wikipedia.org/wiki/Hash_array_mapped_trie
+ */
+export interface AdvancedDirectoryLayout {
+  Data: ByteView<{
+    Type: DataType.HAMTShard,
+    Data: ByteView<Bitfield>
+    /*
+     * HAMT table width (In IPFS it's usually 256)
+     */
+    fanout: uint64,
+    /**
+     * Multihash code for the hashing function used (In IPFS it's [murmur3-64][])
+     *
+     * [murmur3-64]:https://github.com/multiformats/multicodec/blob/master/table.csv#L24
+     */
+    hashType: uint64,
+
+    /**
+     * If omitted to be interpreted as default `0755`. It is RECOMMENDED
+     * to omit if mode matches default.
+     */
+    mode?: Mode
+    mtime?: UnixTime
+  }>
+  Links: NamedLink<FileLayout|DirectoryLayout|DirectoryShard>[]
+}
+
+/**
+ * 
+ */
+export interface DirectoryShard extends PBNode {
+  Data: ByteView<{
+    Type: DataType.HAMTShard,
+    Data: ByteView<Bitfield>
+    /*
+     * HAMT table width (In IPFS it's usually 256)
+     */
+    fanout: uint64,
+    /**
+     * Multihash code for the hashing function used (In IPFS it's [murmur3-64][])
+     *
+     * [murmur3-64]:https://github.com/multiformats/multicodec/blob/master/table.csv#L24
+     */
+    hashType: uint64,
+  }>
+  /**
+   * Either links to other shards or actual directory entries
+   */
+  Links: NamedLink<FileLayout|DirectoryLayout|DirectoryShard>[]
+}
+
+/**
+ * Type for either UnixFS directory representation.
+ */
+export type DirectoryLayout =
+  | FlatDirectoryLayout
+  | AdvancedDirectoryLayout
+
+
+/**
+ * @TODO
+ */
+export type MetadataNode = never
+
+/**
+ * @TODO
+ */
+export interface Symlink extends PBNode {
```

Comment:
Is this accurate ? Given following links I have little confidence 

https://github.com/ipfs/js-ipfs-unixfs/blob/master/packages/ipfs-unixfs/test/unixfs-format.spec.js#L439-L446
https://github.com/ipfs-inactive/js-ipfs-unixfs-engine/issues/3#issuecomment-182440079